macaw-aarch32/src/Data/Macaw/ARM/Identify.hs

```diff
@@ -74,3 +90,122 @@ isReturnValue absProcState val =
+           , MC.ArchSegmentOff ARM.AArch32
+           , ReturnsOnBranch
+           , Seq.Seq (MC.Stmt ARM.AArch32 ids)
+           )
+identifyConditionalReturn mem stmts s finalRegState
+  | not (null stmts)
+  , Just (MC.Mux _ c t f) <- simplifiedMux (s ^. MC.boundValue MC.ip_reg) =
+      case asReturnAddrAndConstant mem finalRegState t f of
+        Just nextIP -> return (c, nextIP, ReturnsOnTrue, stmts)
+        Nothing -> do
+          nextIP <- asReturnAddrAndConstant mem finalRegState f t
+          return (c, nextIP, ReturnsOnFalse, stmts)
+  | otherwise = Nothing
+
+-- | Recognize ARM conditional returns and generate an appropriate arch-specific
+-- terminator
+--
+-- Conditional returns are not supported in core macaw, so we need to use an
+-- arch-specific terminator.  Unlike simple arch-terminators, this one requires
+-- analysis that can only happen in the context of a block classifier.
+--
+-- Note that there are two cases below that could be handled. It seems unlikely
+-- that these would be produced in practice, so they are unhandled for now.
+conditionalReturnClassifier :: MAI.BlockClassifier ARM.AArch32 ids
+conditionalReturnClassifier = do
+  stmts <- CMR.asks MAI.classifierStmts
+  mem <- CMR.asks (MAI.pctxMemory . MAI.classifierParseContext)
+  regs <- CMR.asks MAI.classifierFinalRegState
+  absState <- CMR.asks MAI.classifierAbsState
+  Just (cond, nextIP, returnBranch, stmts') <- return (identifyConditionalReturn mem stmts regs absState)
+  let term = if returnBranch == ReturnsOnTrue then Arch.ReturnIf cond else Arch.ReturnIfNot cond
+  writtenAddrs <- CMR.asks MAI.classifierWrittenAddrs
+
+  jmpBounds <- CMR.asks MAI.classifierJumpBounds
+  ainfo <- CMR.asks (MAI.pctxArchInfo . MAI.classifierParseContext)
+
+  case Jmp.postBranchBounds jmpBounds regs cond of
+    Jmp.BothFeasibleBranch trueJumpState falseJumpState -> do
+      -- Both branches are feasible, but we don't need the "true" case because
+      -- it is actually a return
+      let abs' = MDC.branchBlockState ainfo absState stmts regs cond (returnBranch == ReturnsOnFalse)
+      let fallthroughTarget = ( nextIP
+                              , abs'
+                              , if returnBranch == ReturnsOnTrue then falseJumpState else trueJumpState
+                              )
+      return Parsed.ParsedContents { Parsed.parsedNonterm = F.toList stmts'
+                                   , Parsed.parsedTerm = Parsed.ParsedArchTermStmt term regs (Just nextIP)
+                                   , Parsed.intraJumpTargets = [fallthroughTarget]
+                                   , Parsed.newFunctionAddrs = []
+                                   , Parsed.writtenCodeAddrs = writtenAddrs
+                                   }
+    Jmp.TrueFeasibleBranch _ -> fail "Infeasible false branch"
+    Jmp.FalseFeasibleBranch _ -> fail "Infeasible true branch"
+    Jmp.InfeasibleBranch -> fail "Branch targets are both infeasible"
```

Comment:
Rather than these being a `ClassifyFailure`, shouldn't these be `panic` to match the comment "It seems unlikely... so they are unhandled for now."?

Reply:
I think I'd rather them turn into classify failures rather than hard-stop panics. A user could potentially provide a binary that looked like that, though I'm not concretely sure how.